firedrake/dmplex.pyx

```diff
@@ -1014,3 +964,784 @@ def plex_renumbering(PETSc.DM plex, np.ndarray[PetscInt, ndim=1, mode="c"] reord
+            # to plex edge directions) to get edge directions like below
+            # for each quadrilateral:
+            #
+            #   o-->--o
+            #   |     |
+            #   ^     ^
+            #   |     |
+            #   X-->--o
+            #
+            # Their XOR describes the desired edge directions relative to
+            # the traversal direction of the cone. This is always a
+            # circular permutation of:
+            #
+            #   straight -- straight -- reverse -- reverse
+            #
+            for i in range(4):
+                dst_orient[i] = (cone_orient[i] < 0) ^ facet_orientations[cone[i] - fStart]
+
+            # We select vertex X (figure above) as starting vertex.
+            # Both traversal order (CCW or CW) is fine. We choose the traversal
+            # where the second vertex has the smaller global number.
+            #
+            # The other traversal other would be an equally good choice,
+            # however, for cells in the halo, the same choice must be made in
+            # each MPI process which sees that cell.
+            #
+            # To ensure this, we only calculate cell orientations for the
+            # locally owned cells, and later exchange these values on the
+            # halo cells.
+            if dst_orient[2] and dst_orient[3]:
+                off = 0
+            elif dst_orient[3] and dst_orient[0]:
+                off = 1
+            elif dst_orient[0] and dst_orient[1]:
+                off = 2
+            elif dst_orient[1] and dst_orient[2]:
+                off = 3
+            else:
+                raise RuntimeError("Please get the facet orientation right first!")
+
+            # Cell orientation values are defined to be
+            # the global number of the starting vertex.
+            facet = cone[off]
+
+            CHKERR(DMPlexGetCone(plex.dm, facet, &cone))
+            if cone_orient[off] >= 0:
+                v = cone[0]
+            else:
+                v = cone[1]
+
+            CHKERR(PetscSectionGetOffset(vertex_numbering.sec, v, &V))
+            cell_orientations[cell] = cabs(V)
+
+    return cell_orientations
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+def exchange_cell_orientations(
+    PETSc.DM plex, PETSc.Section section,
+    np.ndarray[np.int32_t, ndim=1, mode="c"] orientations):
+
+    """Halo exchange of cell orientations.
+
+    :arg plex: The DMPlex object encapsulating the mesh topology
+    :arg section: Section describing the cell numbering
+    :arg orientations: Cell orientations to exchange,
+                       values in the halo will be overwritten.
+    """
+    cdef:
+        PETSc.SF sf
+        PetscInt nroots, nleaves
+        PetscInt *ilocal
+        PetscSFNode *iremote
+
+        PETSc.Section new_section
+        np.int32_t *new_values = NULL
+        PetscInt i, c, cStart, cEnd, l, r
+
+    if MPI.comm.size > 1:
```

Comment:
So now we've got globally consistent local (owned) facet orientations and now all we need to do is update the halo region orientations with whatever the owner has, so just do a halo exchange.